vpr/src/place/place.cpp

```diff
@@ -994,26 +1074,86 @@ static void outer_loop_recompute_criticalities(const t_placer_opts& placer_opts,
     prev_inverse_costs->timing_cost = min(1 / costs->timing_cost, MAX_INV_TIMING_COST);
 }
 
-//Update timing information based on current placement by running STA to get new slacks,
-//and calculate updated criticalities and timing costs
-static void recompute_criticalities(float crit_exponent,
-                                    const PlaceDelayModel* delay_model,
-                                    PlacerCriticalities* criticalities,
-                                    ClusteredPinTimingInvalidator* pin_timing_invalidator,
-                                    SetupTimingInfo* timing_info,
-                                    t_placer_costs* costs) {
+static void initialize_timing_info(float crit_exponent,
+                                   const PlaceDelayModel* delay_model,
+                                   PlacerCriticalities* criticalities,
+                                   PlacerSetupSlacks* setup_slacks,
+                                   ClusteredPinTimingInvalidator* pin_timing_invalidator,
+                                   SetupTimingInfo* timing_info,
+                                   t_placer_timing_update_mode* timing_update_mode,
+                                   t_placer_costs* costs) {
+    const auto& cluster_ctx = g_vpr_ctx.clustering();
+    const auto& clb_nlist = cluster_ctx.clb_nlist;
+
+    //Initialize the timing update mode. Update both
+    //setup slacks and criticalities from scratch
+    timing_update_mode->do_update_criticalities = true;
+    timing_update_mode->do_update_setup_slacks = true;
+    timing_update_mode->do_recompute_criticalities = true;
+    timing_update_mode->do_recompute_setup_slacks = true;
+
+    //As a safety measure, for the first time update,
+    //invalidate all timing edges via the pin invalidator
+    //by passing in all the clb sink pins
+    for (ClusterNetId net_id : clb_nlist.nets()) {
+        for (ClusterPinId pin_id : clb_nlist.net_sinks(net_id)) {
+            pin_timing_invalidator->invalidate_connection(pin_id, timing_info);
+        }
+    }
+
+    //Perform timing info update
+    update_setup_slacks_and_criticalities(crit_exponent,
+                                          delay_model,
+                                          criticalities,
+                                          setup_slacks,
+                                          pin_timing_invalidator,
+                                          timing_info,
+                                          timing_update_mode,
+                                          costs);
+
+    //Initialize the data structure that stores committed placer setup slacks
+    commit_setup_slacks(setup_slacks);
+
+    //Don't warn again about unconstrained nodes again during placement
+    timing_info->set_warn_unconstrained(false);
+}
+
+//Update timing information based on current placement by running STA.
```

Comment:
Better to use Doxygen compatible comments for function-level comments which start with /** for a block comment or are /// for a single line.
It's an ongoing project to build more Doxygen docs.